Description: createOutputFile: eliminate stat/open TOCTOU race

From aff4a30807218a52b6b5f200c5aa0eea335547ba Mon Sep 17 00:00:00 2001
From: Kamil Dudka <kdudka@redhat.com>
Date: Mon, 17 Oct 2016 17:59:31 +0200
Subject: [PATCH] createOutputFile: eliminate stat/open TOCTOU race

---
 logrotate.c | 15 ++++++++++-----
 1 file changed, 10 insertions(+), 5 deletions(-)

Original-Author: Kamil Dudka <kdudka@redhat.com>
Origin: https://github.com/logrotate/logrotate/commit/aff4a30807218a52b6b5f200c5aa0eea335547ba
Bug-Ubuntu: http://bugs.launchpad.net/bugs/1630516
Author: Christian Ehrhardt <christian.ehrhardt@canonical.com>
Last-Update: 2017-03-22
Index: logrotate-3.8.7/logrotate.c
===================================================================
--- logrotate-3.8.7.orig/logrotate.c
+++ logrotate-3.8.7/logrotate.c
@@ -299,11 +299,18 @@ static int runScript(struct logInfo *log
 
 int createOutputFile(char *fileName, int flags, struct stat *sb, acl_type acl, int force_mode)
 {
-    int fd;
+    int fd = -1;
     struct stat sb_create;
     int acl_set = 0;
+    int i;
+
+    for (i = 0; i < 2; ++i) {
+	fd = open(fileName, (flags | O_EXCL | O_NOFOLLOW),
+		(S_IRUSR | S_IWUSR) & sb->st_mode);
+
+	if ((fd >= 0) || (errno != EEXIST))
+	    break;
 
-    if (stat(fileName, &sb_create) == 0) {
 	/* the destination file already exists, while it should not */
 	struct tm now = *localtime(&nowSecs);
 	size_t fileName_size = strlen(fileName);
@@ -317,11 +324,9 @@ int createOutputFile(char *fileName, int
 		    fileName, backupName, strerror(errno));
 	    return -1;
 	}
+	/* existing file renamed, try it once again */
     }
 
-	fd = open(fileName, (flags | O_EXCL | O_NOFOLLOW),
-		(S_IRUSR | S_IWUSR) & sb->st_mode);
-
     if (fd < 0) {
 	message(MESS_ERROR, "error creating output file %s: %s\n",
 		fileName, strerror(errno));
